owltorrent-network/src/main/java/edu/rice/owltorrent/network/TorrentManager.java

```diff
@@ -466,20 +493,32 @@ public void handleMessage(PeerMessage message, PeerConnector connector)
               peer.setPeerInterested(false);
               break;
             case HAVE:
+              int idx = ((HaveMessage) message).getIndex();
+              if (!indexToLeechers.containsKey(idx))
+                indexToLeechers.put(idx, ConcurrentHashMap.newKeySet());
+              indexToLeechers.get(idx).add(peer);
               break;
             case BITFIELD:
               Bitfield bitfield = ((BitfieldMessage) message).getBitfield();
               peer.setBitfield(bitfield);
+              boolean leecherFlag = false;
               for (int i = 0; i < torrent.getPieceHashes().size(); i++) {
                 if (!(bitfield.getBit(i))) {
                   if (peers.containsKey(peer)) {
                     leechers.add(peer);
                   }
-                  return;
+                  leecherFlag = true;
+                  break;
                 }
               }
-              if (peers.containsKey(peer)) {
+              if (!leecherFlag && peers.containsKey(peer)) {
                 seeders.add(peer);
+              } else {
+                for (int i = 0; i < torrent.getPieceHashes().size(); i++) {
+                  if (!indexToLeechers.containsKey(i))
+                    indexToLeechers.put(i, ConcurrentHashMap.newKeySet());
```

Comment:
use the concurrent map operation putIfAbsent. (Breaking it into two statements makes it thread unsafe again. )